File: Travel Companion/Model/GeoNames/GeoNamesClient.swift
//
//  GeoNamesClient.swift
//  Travel Companion
//
//  Created by Stefan Jaindl on 17.08.18.
//  Copyright © 2018 Stefan Jaindl. All rights reserved.
//

import Foundation

class GeoNamesClient {
    
    static let sharedInstance = GeoNamesClient()
    
    private init() { }
    
    func fetchCountryCode(latitude: Double, longitude: Double, completionHandler: @escaping (_ errorString: String?, _ result: AnyObject?) -> Void) {
        
        let method = GeoNamesConstants.UrlComponents.path
        
        let queryItems: [String: String] = [GeoNamesConstants.ParameterKeys.latitude: String(latitude),
                                            GeoNamesConstants.ParameterKeys.longitude: String(longitude),
                                            GeoNamesConstants.ParameterKeys.username: SecretConstants.userNameGeoNames]
        
        let url = WebClient.sharedInstance.createUrl(forScheme: GeoNamesConstants.UrlComponents.urlProtocol, forHost: GeoNamesConstants.UrlComponents.domain, forMethod: method, withQueryItems: queryItems)
        
        let request = WebClient.sharedInstance.buildRequest(withUrl: url, withHttpMethod: WebConstants.ParameterKeys.httpGet)
        
        WebClient.sharedInstance.taskForWebRequest(request, errorDomain: "fetchCountryCode", stringResponse: true) { (result, error) in
            
            /* Send the desired value(s) to completion handler */
            if let error = error {
                completionHandler(error.localizedDescription, nil)
            } else {
                completionHandler(nil, result)
            }
        }
    }
}

File: shared/src/commonMain/kotlin/main/model/GeoNames/GeoNamesClient.kt
import com.sjaindl.travelcompanion.SecretConstants
import com.sjaindl.travelcompanion.model.GeoNames.GeoNamesConstants

import io.ktor.client.HttpClient
import io.ktor.client.request.request
import io.ktor.client.response.HttpResponse
import io.ktor.client.response.readText
import io.ktor.http.*
import kotlinx.io.core.use
import kotlinx.serialization.json.Json
import main.model.GeoNames.GeoCodeRepository
import main.model.GeoNames.WebClient

//  GeoNamesClient.swift
//  Travel Companion
//
//  Created by Stefan Jaindl on 17.08.18.
//  Copyright © 2018 Stefan Jaindl. All rights reserved.
//

class GeoNamesClient() {

    companion object {
        val instance = GeoNamesClient()
    }

    val urlComponents = GeoNamesConstants.UrlComponents()
    val endPoint = "${urlComponents.urlProtocol}://${urlComponents.domain}${urlComponents.path}"

    /* // Initialization not possible here because of InvalidMutabilityException on iOS
    private var client = HttpClient {
        install(JsonFeature) {
            serializer = JsonKotlinxSerializer().apply {
                setMapper<Geocode>(Geocode.serializer())
            }
        }
        install(HttpCallValidator)
    }
*/

    suspend fun fetchCountryCode(latitude: Double, longitude: Double, client: HttpClient): GeoCodeRepository = client.request<HttpResponse> {
        method = HttpMethod.Get

        val params = Parameters.build {
            append(GeoNamesConstants.ParameterKeys().latitude, "${latitude}")
            append(GeoNamesConstants.ParameterKeys().longitude, "${longitude}")
            append(GeoNamesConstants.ParameterKeys().username, SecretConstants().userNameGeoNames)
        }.formUrlEncode()

        val url = "${endPoint}?${params}"

        url {
            takeFrom(url)
            //path("lat", latitude.toString(), "lng", longitude.toString(), "username", "jaindl.stefan")
        }

    }.use { response ->

        val json = response.readText()
        WebClient.instance.performBasicWebResponseChecks(response, json)

        val list = Json.nonstrict.parse(GeoCodeRepository.serializer(), json)

        return@use list
    }
}
